packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx

```diff
@@ -83,34 +88,95 @@ export class DetailsHeader extends BaseComponent<IDetailsHeaderProps, IDetailsHe
+    this._setDraggedItemIndex = this._setDraggedItemIndex.bind(this);
+    this._onDragOver = this._onDragOver.bind(this);
+    this._onDrop = this._onDrop.bind(this);
+    this._getHeaderDragDropOptions = this._getHeaderDragDropOptions.bind(this);
+    this._updateDroppingState = this._updateDroppingState.bind(this);
+    this._getDropHintPositions = this._getDropHintPositions.bind(this);
+    this._updateDropHintStates = this._updateDropHintStates.bind(this);
+    this._resetDropHints = this._resetDropHints.bind(this);
+    this._isValidCurrentDropHintIndex = this._isValidCurrentDropHintIndex.bind(this);
+    this._onDropIndexInfo = {
+      sourceIndex: Number.MIN_SAFE_INTEGER,
+      targetIndex: Number.MIN_SAFE_INTEGER
+    };
     this._id = getId('header');
   }
 
   public componentDidMount(): void {
-    const { selection } = this.props;
+    const { selection, columnReorderOptions } = this.props;
     const focusZone = this._root.current;
-    const rootElement = findDOMNode(focusZone as any);
+    const rootElement = findDOMNode(focusZone as any) as HTMLElement;
 
     this._events.on(selection, SELECTION_CHANGE, this._onSelectionChanged);
 
     // We need to use native on this to avoid MarqueeSelection from handling the event before us.
     this._events.on(rootElement, 'mousedown', this._onRootMouseDown);
 
     this._events.on(rootElement, 'keydown', this._onRootKeyDown);
+
+    if (columnReorderOptions) {
+      this._subscriptionObject = this._dragDropHelper!.subscribe(rootElement, this._events, this._getHeaderDragDropOptions());
+    }
+  }
+
+  public componentDidUpdate(): void {
+    const { columnReorderOptions } = this.props;
+    const focusZone = this._root.current;
+    const rootElement = findDOMNode(focusZone as any) as HTMLElement;
+
+    if (!columnReorderOptions) {
+      if (this._subscriptionObject) {
+        this._subscriptionObject.dispose();
+        delete this._subscriptionObject;
+      }
+    } else if (!this._subscriptionObject) {
+      this._subscriptionObject = this._dragDropHelper!.subscribe(rootElement, this._events, this._getHeaderDragDropOptions());
+    }
+
+    if (this._draggedColumnIndex < 0) {
+      this._onDropIndexInfo = {
+        sourceIndex: Number.MIN_SAFE_INTEGER,
+        targetIndex: Number.MIN_SAFE_INTEGER
+      };
+    }
+  }
+
+  public componentWillUpdate(nextProps: IDetailsHeaderProps, nextState: IDetailsHeaderState): void {
+    if (this.props !== nextProps && this._onDropIndexInfo.sourceIndex >= 0 && this._onDropIndexInfo.targetIndex >= 0) {
+      if (this.props.columns[this._onDropIndexInfo.sourceIndex] === nextProps.columns[this._onDropIndexInfo.sourceIndex] && this.props.columns[this._onDropIndexInfo.targetIndex] === nextProps.columns[this._onDropIndexInfo.targetIndex]) {
+        this._onDropIndexInfo = {
+          sourceIndex: Number.MIN_SAFE_INTEGER,
+          targetIndex: Number.MIN_SAFE_INTEGER
+        };
+      }

```

Comment:
I would actually prefer a different condition. Before we re-render, it is important to confirm that onDrop changes has actually been applied. A check for that would be -
this.props.columns[sourceIndex] === newProps.columns[targetIndex]
And if this is not true, we should discard the onDropIndexInfo.